9032 HostnameEndpoint Memory Reactor

markrwilliams opened the pull request:
http://twistedmatrix.com/trac/ticket/9032

This aims to fix a release blocker.

markrwilliams committed:
Ressurect old HostnameEndpoint connect method.

markrwilliams committed:
Existing HostnameEndpoint should implement, not provide.

markrwilliams committed:
Make treq tests pass by avoiding HostnameEndpoint.

exarkun committed:
adapt the reactor instead of the application

markrwilliams committed:
Merge pull request #718 from exarkun/9032-hostnameendpoint-old-memoryreactor

adapt the reactor instead of the application

markrwilliams committed:
Use twisted.logger instead of twisted.python.log

markrwilliams committed:
Wrap a long line.

markrwilliams committed:
Remove _PluggableReactorAdapter by lifting nameResolver onto HostnameEndpoint.

Add deprecation warning.

Fix a bug in _SimpleHostnameResolver._deliver.

Tests for _SimpleHostnameResolver and the revived HostnameEndpoint
behavior.

markrwilliams committed:
Fold HostnameEndpoint._connectWithNameResolver into .connect.

This reduces the overall diff against trunk and eliminates an
opportunity for twistedchecker to complain about missing docstrings.

markrwilliams committed:
Add tests for t.w.client._StandardEndpoint and MemoryReactors.

These tests mostly fail.

markrwilliams committed:
Refactor t.w.client._StandardEndpoint fallbacks and deprecate.

The refactoring allows the deprecation warning to sit in
_StandardEndpoint.__init__, which is one stack frame closer to
Agent.__init__.  It also allows for docstrings that link to a comment
in 9032.

markrwilliams committed:
Topfile for 9032

markrwilliams committed:
Merge branch 'trunk' into 9032-hostnameendpoint-old-memoryreactor

markrwilliams committed:
Methods and URLs are bytes; call Agent.request before deprecation check

markrwilliams committed:
URLs are actually bytes.

markrwilliams committed:
Avoid new deprecation warning by using a compliant reactor.

markrwilliams committed:
Fix method reference.

markrwilliams commented:
There's an unrelated OpenSSL issue that's causing the Windows and macOS builders to fail.

**EDIT** #743 is an attempt at fixing those.

markrwilliams committed:
Merge branch 'trunk' into 9032-hostnameendpoint-old-memoryreactor

markrwilliams committed:
Revert "Add tests for t.w.client._StandardEndpoint and MemoryReactors."

This reverts commit 748f05f429432affb472120a7985bddd2a708bd3.

markrwilliams committed:
Revert "Refactor t.w.client._StandardEndpoint fallbacks and deprecate."

This reverts commit a55f7b9e5d5dd2775a6252b1c58a2a253808c21b.

markrwilliams committed:
Revert "Make treq tests pass by avoiding HostnameEndpoint."

This reverts commit accd508bfc1aa87bc210e14fb7553c3d05d6e780.

markrwilliams committed:
Merge branch 'trunk' into 9032-hostnameendpoint-old-memoryreactor

markrwilliams committed:
Merge branch 'trunk' into 9032-hostnameendpoint-old-memoryreactor

markrwilliams committed:
Merge branch 'trunk' into 9032-hostnameendpoint-old-memoryreactor

markrwilliams committed:
Merge branch 'trunk' into 9032-hostnameendpoint-old-memoryreactor

tomprince commented on src/twisted/internet/endpoints.py:
```diff
@@ -716,6 +809,30 @@ def __init__(self, reactor, host, port, timeout=30, bindAddress=None,
+                reactor.__class__,
+                Version("Twisted", "NEXT", 0, 0),
+                format=("Passing HostnameEndpoint %(fqpn)s"
+                        " was deprecated in %(version)s"),
```
It isn't clear to me that this should be calling out passing a specific type as being deprecated. I think it would be better to simply say passing a reactor that doesn't implement `IReactorPluggableNameResolver` is deprecated. Particularly since the fix might be to make the reactor implement the new interface, rather than changing the reactor being passed.

markrwilliams replied:
The log message should include the interface required, but I'd like it to also include the FQPN of the reactor to make debugging tests easier.

tomprince replied:
Including the FQPN of the reactor seems fine. But I'd go with something like:

> Passing a reactor that doesn't implement IReactorPluggableNameResolver to HostnameEndpoint was deprecated <...>. (`%(fqpn)s` was passed, which doesn't currently implement <...>

that makes it clear that a possible fix is making the reactor implement the given interface.

markrwilliams replied:
I guess it already does mention the interface required of the reactor.

tomprince commented on src/twisted/web/test/test_agent.py:
```diff
@@ -1425,7 +1425,7 @@ def test_deprecatedDuckPolicy(self):
         """
         def warnMe():
-            client.Agent(MemoryReactorClock(),
+            client.Agent(deterministicResolvingReactor(MemoryReactorClock()),
```
Can you file a ticket for either making this public, or extending MemoryReactorClock with this functionality?

markrwilliams replied:
Done!  https://twistedmatrix.com/trac/ticket/9103

glyph replied:
`twisted.testing` HERE WE COME

tomprince commented on src/twisted/topfiles/9032.bugfix:
```diff
@@ -0,0 +1 @@
+twisted.internet.endpoints.HostnameEndpoint and twisted.web.client.Agent work again with reactors that do not provide IReactorPluggableNameResolver.  This undoes the changes that broke downstream users such as treq.testing.  Note that passing reactors that do not provide IReactorPluggableNameResolver to either is deprecated.
```
I think this would be better split into two news files. One for the bugfix, and another one for the deprecation. Also, I don't think it makes sense to call out that this is undoing a previous change: the new behavior is actually different than the old and the point isn't to undo it, but just provide proper deprecation warnings for the change.

tomprince commented on src/twisted/internet/endpoints.py:
```diff
@@ -643,6 +649,89 @@ def _resolvedHostConnect(self, resolvedHost, protocolFactory):
+
+    @ivar _nameResolution: the callable L{resolveHostName} invokes to
+        resolve hostnames.
+    @type _nameResolution: A L{callable} with the same signature as
```
There are some missing words here.

tomprince commented on src/twisted/internet/endpoints.py:
```diff
@@ -643,6 +649,89 @@ def _resolvedHostConnect(self, resolvedHost, protocolFactory):
+        resolutionReceiver.resolutionBegan(HostResolution(hostName))
+        d = self._nameResolution(hostName, portNumber)
+        d.addCallback(self._deliver, resolutionReceiver)
+        d.addErrback(self._ebLog, hostName)
```
I don't think there is any value in having these callbacks out-of-line. There is a bunch of code like this in twisted, but I *think* that newer code tends to avoid it.

markrwilliams replied:
It does - I took this approach because I had pretensions of testing each of these methods.  But coverage is good, so I might as well switch back to closures.

markrwilliams replied:
Also the docstrings looked nicer on instance methods than they did on closures.

tomprince replied:
If they are closures, they might not need docstrings (certainly they could do with more abbreviated docstrings in that case).

markrwilliams replied:
They don't - I found them useful while writing the patch, so I thought I'd leave them in.

tomprince commented on src/twisted/internet/endpoints.py:
```diff
@@ -888,6 +1016,22 @@ def cancelRemainingPending(result):
+    def _nameResolution(self, host, port):
+        """
+        Resolve the hostname string into a tuple containing the host
+        address.
```
It would probably be good to make it clear at least in the docstring and perhaps in the method name that this method is just used as a fallback when the reactor provided to the endpoint doesn't implement the new interface.

tomprince commented on src/twisted/internet/endpoints.py:
```diff
@@ -888,6 +1016,22 @@ def cancelRemainingPending(result):
+            return value.
+        """
+        return self._deferToThread(self._getaddrinfo, host, port, 0,
+                socket.SOCK_STREAM)
```
I'm not sure that it makes sense to allow overriding `deferToThread` and `getaddrinfo` individually, instead of just overriding this function.

markrwilliams replied:
`deferToThread` has to be replaceable to make this function testable.  The original implementation also made `getaddrinfo` replaceable, apparently with the intent of verifying that `deferToThread` invokes it.

I was inclined to leave these two staticmethods in for two reasons:

1. My tests looked more like the existing ones, apparently verifying the same behaviors.
2.  A vague concern that some tests, inside or outside Twisted itself, relied on it.  

A cursory GitHub search indicated that 2. wasn't the case.  1. remains true and useful, I think.

tomprince replied:
Looking at the tests, everywhere that `._deferToThread` and `._getaddrinfo` are overridden could be replaced by

```
self._nameResolution = fakegetaddrinfo
```

(modulo fakegetaddrinfo taking some additional arguments that it ignores). That would mean that `_nameResolution` isn't covered by the tests but I don't think the coverage that it currently gets is meaningful, since in tests it is just calling fake function that aren't verified to be equivalent to the real thing anyway.

markrwilliams replied:
@tomprince I've filed [a new ticket](https://twistedmatrix.com/trac/ticket/9147) that we can use to track this improvement.

tomprince commented on src/twisted/internet/test/test_endpoints.py:
```diff
@@ -1687,6 +1688,381 @@ class WithResolver(proxyForInterface(
+
+    def connectArgs(self):
+        """
+
```
There is an extra newline here.

markrwilliams replied:
`twistedchecker` complains if there isn't a newline here, and it complains if the return type isn't documented.

tomprince replied:
That seems like a bug in twistedchecker. Or an indication there should be some prose here.

markrwilliams replied:
I think it's a bug in twistedchecker.  All of the other test classes that define `connectArgs` have the same docstring, minus the newline, presumably to quiet twistedchecker.

glyph replied:
Please file & link the bug in twistedchecker so we know about it, per policy :)

markrwilliams replied:
Looks like @rodrigc filed the bug: https://github.com/twisted/twistedchecker/issues/124

tomprince reviewed (changes requested):
This looks like a good improvement. There are a couple of small changes, and I didn't have time to look at the tests.

markrwilliams committed:
Merge branch 'trunk' into 9032-hostnameendpoint-old-memoryreactor

markrwilliams committed:
Address review feedback.

See https://github.com/twisted/twisted/pull/740#pullrequestreview-30161333

markrwilliams committed:
Test _fallbackNameResolution explicitly.

markrwilliams committed:
IP addresses are text on Python 3.

markrwilliams committed:
Don't check the protocol in getaddrinfo.

Windows leaves this as AF_UNSPEC if you don't specify it, and we
shouldn't test against it anyway in case quic or something makes it
into a kernel.

markrwilliams committed:
Merge branch 'trunk' into 9032-hostnameendpoint-old-memoryreactor

markrwilliams committed:
topfile -> newsfragments

glyph commented on src/twisted/internet/test/test_endpoints.py:
```diff
@@ -1687,6 +1688,412 @@ class WithResolver(proxyForInterface(
+
+class HostnameEndpointFallbackNameResolutionTests(unittest.TestCase):
+    """
+    Verify that L{HostnameEndpoint._fallbackNameResolution} defers a
```
"verify that" -> https://jml.io/pages/test-docstrings.html

glyph commented on src/twisted/internet/test/test_endpoints.py:
```diff
@@ -1687,6 +1688,412 @@ class WithResolver(proxyForInterface(
+
+    def test_fallbackNameResolution(self):
+        """
+        A L{Deferred} is returned that fires with the resoution of the
```
Active voice not passive voice, "`_fallbackNameResolution` returns a `Deferred` that..."

glyph commented on src/twisted/internet/test/test_endpoints.py:
```diff
@@ -1687,6 +1688,412 @@ class WithResolver(proxyForInterface(
+
+    def test_endpointConnectFailure(self):
+        """
+        If L{HostnameEndpoint.connect} is invoked and there is no server
```
"when" is preferable to "if", because it stipulates the existence of a supported configuration.

glyph reviewed (approved):
Some minor observations inline, but pending a clean CI run I think this is more than baked.  Please address and land.

markrwilliams committed:
Improve docstrings per @glyph's review.  Thanks!

(see https://github.com/twisted/twisted/pull/740#pullrequestreview-39834618)

markrwilliams commented:
The failing Codecov status appears to be down to partial coverage on [this line](https://codecov.io/gh/twisted/twisted/pull/740/diff#D2-705).  That could be covered by resolving an IPv6 address. 

We now [skip tests that need IPv6 on platforms that don't support it](http://twistedmatrix.com/trac/ticket/9144), but the way we determine that [needs to be generalized](https://twistedmatrix.com/trac/ticket/9146#ticket).  Rather than introduce another instance of this function's definition in this patch, I'll merge with the slightly reduced coverage and address it in the patch for [9146](https://twistedmatrix.com/trac/ticket/9146#ticket).
